Sandbox wowoto

wowoto9772 opened the pull request:
sandbox/sandbox.py 내 

Sandbox 생성자에 호스트와 컨테이너 사이에 포트 바인딩 추가

컨테이너에 파일 및 바이트 어레이를 복사하는 메소드 추가 및 

tests/test_sandbox.py 내

테스트 코드 추가

tests/resources 내

server.py, client.py 코드를 추가하였는데,

호스트와 컨테이너 사이에 이미 바인딩한 포트를 사용해서 tcp communication을 한다.

wowoto9772 committed:
add copy file(bytes) to container and their tests

- two functions are returned {Success, Bytes} data.

wowoto9772 committed:
in travis settings, edit before install

- images repository is changed.

wowoto9772 committed:
in sandbox/sandbox.py, edit finding absolute path

- 'locate' is not appropriate.
- os.path.dirname(__file__) return current file's absolute path

wowoto9772 committed:
Merge remote-tracking branch 'origin/sandbox' into sandbox/wowoto

nona1314 commented on sandbox/sandbox.py:
```diff
@@ -1,30 +1,86 @@
+        cur_directory = os.path.dirname(__file__)  # find absolute path
+        os.chdir(cur_directory)
+
+        data_tar_stream = BytesIO()
```
이 섹션(https://docs.python.org/3/library/io.html#in-memory-streams)에 BytesIO 언급된것 처럼 Tar을 In memory로 만드려고 한것 같은데 좀 단계별로 주석을 더 달아도 되지 않을까요?

wowoto9772 replied:
ok

nona1314 commented on sandbox/sandbox.py:
```diff
@@ -1,30 +1,86 @@
+        os.chdir(cur_directory)
+
+        data_tar_stream = BytesIO()
+        data_tar = tarfile.TarFile(fileobj=data_tar_stream, mode="w")
```
from https://docs.python.org/3/library/tarfile.html
> class tarfile.TarFile
> Class for reading and writing tar archives. **Do not use this class directly**: use tarfile.open() instead. See TarFile Objects.

open으로 바꾸어 주세요.

wowoto9772 replied:
ok

nona1314 commented on sandbox/sandbox.py:
```diff
@@ -1,30 +1,86 @@
+
+        tarinfo = tarfile.TarInfo(name="server.py")
+        tarinfo.size = len(data)
+        tarinfo.mtime = time.time()
```
size하고 name만 있으면 안되나요? mtime도 옮겨줘야 하나요?

wowoto9772 replied:
mtime을 통해서, 컨테이너로 전송된 시점을 기록하고 싶었던 건데,

불필요한 부분이라고 생각하면 상의해서 제거해도 될 듯.

nona1314 commented on sandbox/sandbox.py:
```diff
@@ -1,30 +1,86 @@
+            data=data_tar_stream
+        ), "Bytes": data}
+
+    def copy_bytes_to_container(self, bytes_data: bytes, container_directory: str, file_name: str) -> dict:
```
`copy_file_to_container` 하고 상당부분 겹치는데 하나로 만들 수는 없을까요

wowoto9772 replied:
bytes를 살릴지 file을 살릴지가 고민이었는데, 결국은 bytes로 바꿔서 전송하니까

file을 없에고 bytes만 살리는 방향으로 하면 될 것 같아.

nona1314 commented:
docker remote api에 보면 get_archive 가 tar stream을 반환한다는데
```py
data = []
tar_stream = BytesIO(tar_stream_from_container)
tarobj = tarfile.open(fileobj=tar_stream)
for tarinfo in tarobj:
    bufio = tarobj.extractfile(tarinfo)
    data.append(bufio.read())
```

이런식으로 container to host copy도 만들 수 있을 것 같은데..?

주석 약간이랑 변수명, 위에 인라인 커멘트, container to host copy 정도 더해도 될거 같아요.

wowoto9772 committed:
add comment and edit method, parameter name

- change parameter 'data' into 'source'
- tarfile.Tarfile -> tarfile.open
- add explanation step by step

wowoto9772 committed:
edit method, and add test

- method name change
copy_file_to_container -> copy_host_file_to_container
copy_bytes_to_container -> copy_host_to_container

- add test host file not found

wowoto9772 committed:
find file name using in pathlib PurePosixPath
